File: 03-creational-patterns/04-lazy-initialization.go
package main

import (
	"fmt"
	"sync"
)

type LazyResource struct {
	data []int
	once sync.Once
}

func (lr *LazyResource) GetData() []int {
	lr.once.Do(func() {
		fmt.Println("Initializing data...")
		lr.data = make([]int, 1000000)
		for i := range lr.data {
			lr.data[i] = i * i
		}
	})
	return lr.data
}

func main() {
	resource := &LazyResource{}
	fmt.Println("Before accessing data")
	fmt.Println(resource.GetData()[:5]) // Data gets initialized here
	fmt.Println("GetData() again....")
	fmt.Println(resource.GetData()[:5]) // Data gets initialized here
}

File: 03-creational-patterns/04-lazy-initialization.py
class LazyResource:
    def __init__(self):
        self._data = None

    @property
    def data(self):
        if self._data is None:
            print("Initializing data...")
            self._data = self._expensive_operation()
        return self._data

    def _expensive_operation(self):
        return [i * i for i in range(1000000)]

# Usage
resource = LazyResource()
print("Before accessing data")
print(resource.data[:5])  # Data gets initialized here